### PR TITLE
[Mistral] Mistral-7B-v0.1 support

### PR DESCRIPTION
**Pull request by Bam4d**
# What does this PR do?

Support for Mistral 7B models 

Fixes # (issue)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: docs/source/en/model_doc/mistral.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+Mistral-7B-v0.1 is Mistral AI’s first Large Language Model (LLM). 
+
+## Model Details
+
+Mistral-7B-v0.1 is a decoder-based LM with the following architectural choices:
+* Sliding Window Attention - Trained with 8k context length and fixed cache size, with a theoretical attention span of 128K tokens
+* GQA (Grouped Query Attention) - allowing faster inference and lower cache size.
+* Byte-fallback BPE tokenizer - ensures that characters are never mapped to out of vocabulary tokens.
+
+We also provide an instruction fine-tuned model: `Mistral-7B-Instruct-v0.1` which can be used for chat-based inference.
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/mistral.md:30]: Would be awesome to have more details about the model directly here like the data it was trained on, but it's a nit!

File: src/transformers/models/mistral/modeling_mistral.py [hunk 2]
```diff
@@ -0,0 +1,933 @@
...
+    batch, num_key_value_heads, slen, head_dim = hidden_states.shape
+    if n_rep == 1:
+        return hidden_states
+    hidden_states = hidden_states[:, :, None, :, :].expand(batch, num_key_value_heads, n_rep, slen, head_dim)
+    return hidden_states.reshape(batch, num_key_value_heads * n_rep, slen, head_dim)
+
+
+class MistralAttention(nn.Module):
+    """
+    Multi-headed attention from 'Attention Is All You Need' paper. Modified to use sliding window attention
```
Comment (ArthurZucker) [inline src/transformers/models/mistral/modeling_mistral.py:187]: Last nit, is there a link to a doc/article/blog where the window attention is introduced? 🤗

**Review by ArthurZucker — APPROVED** [review]
Just left 2 nits, thanks a lot for opening the PR and adding this model to transformers! 
Regarding the CIs, `pip install -e ".[quality"` should make sure you have the correct versioning to run  `make style` and `make fixup`. The hub test is flaky

**Comment by ArthurZucker** [comment]
Failing test is expected, you can add the `configuration_mistral` file to the `not_doctested.txt` file. 
The error: 
```python 
failed on setup with "worker 'gw0' crashed while running 'src/transformers/models/mistral/configuration_mistral.py::transformers.models.mistral.configuration_mistral.MistralConfig'"
worker 'gw0' crashed while running 'src/transformers/models/mistral/configuration_mistral.py::transformers.models.mistral.configuration_mistral.MistralConfig'
``` 
is just from trying to init a too big model for the runner! (Llama also had this issue)

File: src/transformers/models/mistral/modeling_mistral.py (hunk 2 above)
Comment (timlacroix) [reply src/transformers/models/mistral/modeling_mistral.py]: Done :)

**Pull request merged**